File: src/leetcode/python/AddAndSearchWord_211.py
class Dict(object):
    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.is_word = False
        self.dicts = {}


class AddAndSearchWord_211(object):
    def __init__(self):
        """
        initialize your data structure here.
        """
        self.root = Dict()

    def addWord(self, word):
        """
        Adds a word into the data structure.
        :type word: str
        :rtype: void
        """
        node = self.root
        for c in word:
            if c not in node.children:
                node.children[c] = Dict()
            node = node.children[c]
        node.is_word = True

    def search(self, word):
        """
        Returns if the word is in the data structure. A word could
        contain the dot character '.' to represent any one letter.
        :type word: str
        :rtype: bool
        """
        def nextSearch(n, j):
            for nd in n.dicts:
                k = j
                while k < len(word):
                    wd = word[k]
                    if wd == '.':
                        if nextSearch(nd, k+1):
                            return True
                        else:
                            break
                    else:
                        if wd not in nd.dicts:
                            break
                        nd = nd.dicts[wd]
                    k += 1
                if k == len(word) and nd.is_word:
                    return True
            return False

        node = self.root
        for i in range(len(word)):
            w = word[i]
            if w == '.':
                return nextSearch(node, i + 1)
            else:
                if w not in node.dicts:
                    return False
                node = node.dicts[w]
        return node.is_word

File: src/leetcode/java/AddAndSearchWord_211.java
package leetcode.java.medium;

public class AddAndSearchWord_211 {
  private Dict root = new Dict();

  // Adds a word into the data structure.
  public void addWord(String word) {
    Dict pt = root;
    int idx;
    for (char c : word.toCharArray()) {
      idx = c - 'a';
      if (pt.dicts[idx] == null) pt.dicts[idx] = new Dict();
      pt = pt.dicts[idx];
    }
    pt.isWord = true;
  }

  // Returns if the word is in the data structure. A word could
  // contain the dot character '.' to represent any one letter.
  public boolean search(String word) {
    char[] wc = word.toCharArray();
    Dict pt = root;
    for (int i = 0; i < wc.length; i++) {
      if (wc[i] == '.') {
        return nextSearch(pt, wc, i + 1);
      } else {
        int index = wc[i] - 'a';
        if (pt.dicts[index] == null) return false;
        pt = pt.dicts[index];
      }
    }
    return pt.isWord;
  }

  private boolean nextSearch(Dict pointer, char[] wc, int start) {
    Dict p = null;
    for (int j = 0; j < pointer.dicts.length; j++) {
      p = pointer.dicts[j];
      if (p != null) {
        int i;
        for (i = start; i < wc.length; i++) {
          if (wc[i] == '.') {
            if (nextSearch(p, wc, i + 1)) return true;
            else break;
          } else {
            int index = wc[i] - 'a';
            if (p.dicts[index] == null) break;
            p = p.dicts[index];
          }
        }
        if (i == wc.length && p.isWord) return true;
      }
    }
    return false;
  }

  class Dict {
    Dict[] dicts;
    boolean isWord;

    Dict() {
      this.dicts = new Dict[26];
      this.isWord = false;
    }
  }
}
